
In Cross-compile, we need to regenerate -- if needed -- using a native-built
Python.  This patch includes additional vebose output and fail-fast logic for
when the PGEN determination is incorrectly done to avoid the 55-minute build
that ultimately fails

--- configure.ac.orig	2016-06-25 23:38:39.000000000 +0200
+++ configure.ac	2016-12-05 22:06:40.183563868 +0100
@@ -74,6 +74,8 @@
 	fi
         AC_MSG_RESULT($interp)
 	PYTHON_FOR_BUILD='_PYTHON_PROJECT_BASE=$(abs_builddir) _PYTHON_HOST_PLATFORM=$(_PYTHON_HOST_PLATFORM) PYTHONPATH=$(shell test -f pybuilddir.txt && echo $(abs_builddir)/`cat pybuilddir.txt`:)$(srcdir)/Lib _PYTHON_SYSCONFIGDATA_NAME=_sysconfigdata_$(ABIFLAGS)_$(MACHDEP)_$(MULTIARCH) '$interp
+	else
+        AC_MSG_RESULT($PYTHON_FOR_BUILD)
     fi
 elif test "$cross_compiling" = maybe; then
     AC_MSG_ERROR([Cross compiling required --host=HOST-TUPLE and --build=ARCH])
@@ -82,6 +84,19 @@
 fi
 AC_SUBST(PYTHON_FOR_BUILD)
 
+if test "$cross_compiling" = yes; then
+    AC_MSG_CHECKING([for pgen for cross build])
+    if test -z "$PGEN_FOR_BUILD"; then
+	    which pgen >/dev/null 2>&1 || AC_MSG_ERROR([pgen not found])
+	    PGEN_FOR_BUILD=pgen
+	fi
+    AC_MSG_RESULT($PGEN_FOR_BUILD)
+else
+    PGEN_FOR_BUILD='$(PGEN)'
+fi
+AC_SUBST(PGEN_FOR_BUILD)
+
+
 dnl Ensure that if prefix is specified, it does not end in a slash. If
 dnl it does, we get path names containing '//' which is both ugly and
 dnl can cause trouble.
--- Makefile.pre.in.orig	2016-06-25 23:38:37.000000000 +0200
+++ Makefile.pre.in	2016-12-05 22:08:58.803564372 +0100
@@ -286,6 +286,7 @@
 ##########################################################################
 # Parser
 PGEN=		Parser/pgen$(EXE)
+PGEN_FOR_BUILD=	@PGEN_FOR_BUILD@
 
 PSRCS=		\
 		Parser/acceler.c \
@@ -786,11 +786,11 @@
 		$(CC) $(OPT) $(PY_CORE_LDFLAGS) $(PGENOBJS) $(LIBS) -o $(PGEN)
 
 .PHONY: regen-grammar
-regen-grammar: $(PGEN)
+regen-grammar: $(PGEN_FOR_BUILD)
 	# Regenerate Include/graminit.h and Python/graminit.c
 	# from Grammar/Grammar using pgen
 	@$(MKDIR_P) Include
-	$(PGEN) $(srcdir)/Grammar/Grammar \
+	$(PGEN_FOR_BUILD) $(srcdir)/Grammar/Grammar \
 		$(srcdir)/Include/graminit.h.new \
 		$(srcdir)/Python/graminit.c.new
 		$(UPDATE_FILE) $(srcdir)/Include/graminit.h $(srcdir)/Include/graminit.h.new
